### PR TITLE
Remove ArrayClass::m_ElementType - get type from the type handle on MethodTable

### PR DESCRIPTION
**Pull request by elinor-fung**
`ArrayClass::m_ElementType` held a cached version of `MethodTable::m_ElementTypeHnd.GetSignatureCorElementType()`. This removes that member and makes `MethodTable::GetArrayElementType` (and therefore `ArrayBase::GetArrayElementType`) use the type handle to get the element type instead.

This also fixes an issue where we were not always emitting a type check in array functions on arrays with elements that were object references. The added tests validate that these cases now result in an `ArrayTypeMismatchException` for mismatched element types.

https://github.com/dotnet/runtime/pull/106718#discussion_r1725220897

cc @jkotas @davidwrighton @AaronRobinsonMSFT

### CODE REVIEW

**Review by AaronRobinsonMSFT — COMMENTED** [review]
This looks fine to me, but I defer to @jkotas or @davidwrighton for the final sign off. There is too much history here.

File: src/coreclr/vm/array.cpp [hunk 1]
```diff
@@ -658,9 +657,8 @@ class ArrayOpLinker : public ILStubLinker
             hiddenArgIdx = 0;
         }
 #endif
-
-        ArrayClass *pcls = (ArrayClass*)(pMT->GetClass());
-        if(pcls->GetArrayElementType() == ELEMENT_TYPE_CLASS)
+        CorElementType sigElementType = pMT->GetArrayElementType();
+        if(sigElementType == ELEMENT_TYPE_CLASS || sigElementType == ELEMENT_TYPE_ARRAY)
```
Comment (jkotas) [inline src/coreclr/vm/array.cpp:661]: Is this fixing a user observable bug that we should add a regression case for?

**Comment by jkotas** [comment]
I reviewed places where we are using `GetArrayElementType()` currently. The new implementation of `GetArrayElementType` is somewhat slower compared to what we have today. It may be a good to change `pArrayMT->GetArrayElementType() == ELEMENT_TYPE_R8` to `pArrayMT->GetArrayElementTypeHandle() == CoreLibBinder::GetElementType(ELEMENT_TYPE_R8)` in gchelpers.cpp to avoid adding path length to the gc alloc path on x86. The rest looks fine.

File: src/coreclr/vm/array.cpp (hunk 1 above)
Comment (jkotas) [reply src/coreclr/vm/array.cpp:661]: Try something like this:
```
using System;

var a = new B[1][];

var d = typeof(object[][]).GetMethod("Set").CreateDelegate<Action<object[][], int, object[]>>();
d(a, 0, new object[] { new object() });

B actuallyObject = a[0][0];
Console.WriteLine(actuallyObject.M());

class B
{
    public virtual string M() => "Hello world";
}
```

Comment (jkotas) [reply src/coreclr/vm/array.cpp:661]: It crashes with fatal error on .NET 9.

Comment (jkotas) [reply src/coreclr/vm/array.cpp:661]: I think this condition should be `if (CorTypeInfo::IsObjRef(sigElementType))`

**Review by jkotas — APPROVED** [review]

**Review by davidwrighton — APPROVED** [review]

File: src/tests/baseservices/invalid_operations/Arrays.cs [hunk 2]
```diff
@@ -0,0 +1,51 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Reflection;
+
+using Xunit;
+
+[ConditionalClass(typeof(TestLibrary.PlatformDetection), nameof(TestLibrary.PlatformDetection.IsMonoInterpreter))]
```
Comment (jkotas) [inline src/tests/baseservices/invalid_operations/Arrays.cs:9]: Should this be disabled against a bug?

**Pull request merged**